src/material_library.cpp

```diff
@@ -0,0 +1,352 @@
+  material_library.erase(key);
+  keylist.erase(key);
+}
+
+pyne::Material pyne::MaterialLibrary::get_material(
+    const std::string& mat_name) const {
+  return *(this->get_material_ptr(mat_name));
+}
+
+pyne::shr_mat_ptr pyne::MaterialLibrary::get_material_ptr(
+    const std::string& mat_name) const {
+  auto it = material_library.find(mat_name);
+  if (it != material_library.end()) {
+    return it->second;
+  } else {
+    return std::make_shared<pyne::Material>(pyne::Material());
+  }
+}
+
+void pyne::MaterialLibrary::write_hdf5(const std::string& filename,
+                                       const std::string& datapath) const {
+  // Turn off annoying HDF5 errors
+  H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+  // Set file access properties so it closes cleanly
+  hid_t fapl;
+  fapl = H5Pcreate(H5P_FILE_ACCESS);
+  H5Pset_fclose_degree(fapl, H5F_CLOSE_STRONG);
+  hid_t matlib_grp_id;
+  hid_t data_id;
+  hid_t db;
+  if (!pyne::file_exists(filename)) {
+    // Create the file
+    db = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, fapl);
+  } else {
+    db = H5Fopen(filename.c_str(), H5F_ACC_RDWR, fapl);
+  }
+  // Check if root_path exist and what type it is
+  std::string root_path = "/material_library";
+
+  H5O_info_t object_info;
+  herr_t status = H5Eset_auto2(H5E_DEFAULT, NULL, NULL);
+
+  // Check if root_path exist and what type it is
+  status =
+      H5Oget_info_by_name(db, root_path.c_str(), &object_info, H5P_DEFAULT);
+  if (status == 0) {
+    // "/material_library" not a group: fail!
+    if (object_info.type != H5O_TYPE_GROUP) {
+      throw std::runtime_error(
+          "Non-group/non-dataset object /material_library already exists in "
+          "the file. Can't write the Material");
+    } else {  // "/material_library" is a group get its hid
+      matlib_grp_id = H5Gopen2(db, root_path.c_str(), H5P_DEFAULT);
+    }
+  } else {  // "/material_library" does not exist -> create it !
+    matlib_grp_id = H5Gcreate2(db, root_path.c_str(), H5P_DEFAULT, H5P_DEFAULT,
+                               H5P_DEFAULT);
+  }
+
+  // Group "/root_path/datapath" does not exist create it
+  std::string full_path = root_path + datapath;
+  if (!h5wrap::path_exists(db, full_path)) {
+    data_id = H5Gcreate2(db, full_path.c_str(), H5P_DEFAULT, H5P_DEFAULT,
+                         H5P_DEFAULT);
+  } else {
+    data_id = H5Gopen2(db, full_path.c_str(), H5P_DEFAULT);
+  }
+  std::string compath = "composition";
+  std::string nucpath = "nuclidelist";
+  write_hdf5_nucpath(data_id, nucpath);
+  std::vector<int> nuc_list;
+  nuc_list.assign(nuclist.begin(), nuclist.end());
+  for (auto mat : material_library) {
+    mat.second->write_hdf5_datapath(data_id, compath, -0.0,
+                                    DEFAULT_MAT_CHUNKSIZE, nuc_list);
+  }
+
+  H5Fflush(db, H5F_SCOPE_GLOBAL);
+  H5Gclose(data_id);
+  H5Gclose(matlib_grp_id);
+  H5Fclose(db);
+}
+
+void pyne::MaterialLibrary::write_hdf5_nucpath(hid_t db,
```

Comment:
I didn't see a response about reusing the method in `Material` for this?

Related: how do we deal with overwriting a material library that already exists?  If the nucpath is already there, does this work?

Reply:
the 2 methods are "very" different, the goal here is to overwrite the exciting nuclist, where in `Material` we want to reused the existing one.

I need to add the test to make sure this is actually working as I expect.

Reply:
so last week, I did some testing, did't go as smoothly as I wanted.
Basically the planned behavior was to overwrite the existing material library if trying to write on the same datapath.
Did't work, I ended up messing up with nuclides list....

As HDF5 does not support (or at least was not able to find it) overwrite, I think we are stuck with 2 options, when trying to overwrite in a `datapath` :
- check if the hdf5 nuclide list is compatible with the material library we are trying to write and just append
- miserably fail and whine about it.

the first option looks scary because we will enlarge a material library we don't know anything about....

Reply:
Maybe H5Ldelete (https://www.mathworks.com/matlabcentral/answers/395920-how-can-i-delete-a-dataset-completely-from-a-group-in-a-hdf5-file#answer_356519)?

Reply:
I forgot about this option, 
As far as I remember, it is not possible to recover the space of the deleted table, but it might the the lesser of 2 evils....
I'll take a deeper look.

Reply:
there is h5unlink which would 'loose' the old data allowing you to write new

Reply:
We will lose the file space, but a MaterialLibrary is generally pretty small compared to most files, so this shouldn't matter much.

Reply:
last question, shall we implement the overwriting behavior by default or add a bool to enable it (disabling overwriting by default) and warn the user if the capability exists (using the bool ?)

Reply:
I guess it's always nicest to the user to prevent them from causing damage :)  Let's add a boolean to enable it, and issue an error if it already exists.

Once this is in, we can also add tools to extract a MaterialLibrary so that a user can save the old one, and then force overwriting it in a file.

Reply:
All the elements of such tool are already there, as we can load a hdf5 datapath into a MaterialLibrary and write it into the targeted path :)

Just a matter of putting all the pieces together.